### PR TITLE
CSHARP-3757: Redirect read/write retries to other mongos if possible

### PR DESCRIPTION
**Pull request by adelinowona**
### Description
On a sharded cluster, retrying reads and writes will attempt to prioritize other mongos for selection if available.

### What is changing?
- Server selection now takes a collection of servers to deprioritize and will attempt to exclude those servers when the topology is sharded and multiple mongoses are present.
- Adds unit tests for server selection and prose tests for retryable reads and writes.
- Adds functionality to provide a collection of servers to deprioritize in relevant places.

### CODE REVIEW

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs [hunk 1]
```diff
@@ -413,6 +413,101 @@ public void SelectServer_should_throw_if_any_servers_are_incompatible(int min, i
...
+            var connected1 = ServerDescriptionHelper.Connected(subject.Description.ClusterId, endPoint1);
+            var connected2 = ServerDescriptionHelper.Connected(subject.Description.ClusterId, endPoint2);
+            var connected3 = ServerDescriptionHelper.Connected(subject.Description.ClusterId, endPoint3);
+            subject.SetServerDescriptions(connected1, connected2, connected3);
+
+            var selector = new DelegateServerSelector((c, s) => s);
+
+            var deprioritizedServers = new List<ServerDescription> { connected1 };
+
+            for (int i = 0; i < 15; i++)
```
Comment (adelinowona) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs:440]: Running multiple times to account for server selection randomness and guarantee we get the expected result for each run.

File: src/MongoDB.Driver.Core/Core/Bindings/IBinding.cs [hunk 2]
```diff
@@ -61,6 +62,22 @@ public interface IReadBinding : IBinding
...
         /// <returns>A channel source.</returns>
         Task<IChannelSourceHandle> GetReadChannelSourceAsync(CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Gets a channel source for read operations while deprioritizing servers in the provided collection.
+        /// </summary>
+        /// <param name="deprioritizedServers">The deprioritized servers.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>A channel source.</returns>
+        IChannelSourceHandle GetReadChannelSource(IReadOnlyCollection<ServerDescription> deprioritizedServers, CancellationToken cancellationToken);
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Bindings/IBinding.cs:72]: This interface is pseudo public (going to be hidden in 3.0), and we are introducing a breaking change anyway.
So should we consider just adding a parameter to the existing method?

File: src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs [hunk 3]
```diff
@@ -506,13 +516,41 @@ public IServer SelectServer()
                 _connectedServers.Clear();
                 _connectedServerDescriptions.Clear();
 
+                var excludingDeprioritizedServers = deprioritizedServers != null && _cluster.Description.Type == ClusterType.Sharded;
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Clusters/Cluster.cs:519]: Consider micro optimization: `deprioritizedServers?.Any() == true`

File: src/MongoDB.Driver.Core/Core/Operations/RetryableReadOperationExecutor.cs [hunk 4]
```diff
@@ -51,7 +51,7 @@ public static TResult Execute<TResult>(IRetryableReadOperation<TResult> operatio
 
             try
             {
-                context.ReplaceChannelSource(context.Binding.GetReadChannelSource(cancellationToken));
+                context.ReplaceChannelSource(context.Binding.GetReadChannelSource(new []{ context.ChannelSource.ServerDescription }, cancellationToken));
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Operations/RetryableReadOperationExecutor.cs:54]: minor: missing space `new [] {}`

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs [hunk 5]
```diff
@@ -413,6 +413,101 @@ public void SelectServer_should_throw_if_any_servers_are_incompatible(int min, i
...
+                _capturedEvents.Next().Should().BeOfType<ClusterSelectedServerEvent>();
+                _capturedEvents.Any().Should().BeFalse();
+            }
+        }
+
+        [Theory]
+        [ParameterAttributeData]
+        public void SelectServer_should_return_deprioritized_servers_if_no_other_servers_exist_or_cluster_not_sharded(
+            [Values(false, true)] bool async,
+            [Values(false, true)] bool IsSharded)
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs:469]: minor: lower case `isSharded`

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs [hunk 6]
```diff
@@ -413,6 +413,101 @@ public void SelectServer_should_throw_if_any_servers_are_incompatible(int min, i
...
+            [Values(false, true)]
+            bool async)
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            var subject = CreateSubject(ClusterConnectionMode.Sharded);
+#pragma warning restore CS0618 // Type or member is obsolete
+
+            subject.Initialize();
+
+            var endPoint1 = new DnsEndPoint("localhost", 27017);
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs:428]: minor: add simple test to validate `deprioritizedServers == null ` case

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs [hunk 7]
```diff
@@ -413,6 +413,101 @@ public void SelectServer_should_throw_if_any_servers_are_incompatible(int min, i
...
+            var deprioritizedServers = new List<ServerDescription> { connected1 };
+
+            for (int i = 0; i < 15; i++)
+            {
+                _capturedEvents.Clear();
+
+                IServer result;
+                if (async)
+                {
+                    result = subject.SelectServerAsync(selector, deprioritizedServers, CancellationToken.None).GetAwaiter().GetResult();
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs:447]: Make the test return `Task`, and just await here.

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs [hunk 8]
```diff
@@ -413,6 +413,101 @@ public void SelectServer_should_throw_if_any_servers_are_incompatible(int min, i
...
+                }
+                else
+                {
+                    result = subject.SelectServer(selector, deprioritizedServers, CancellationToken.None);
+                }
+
+                result.Should().NotBeNull();
+
+                var deprioritizedServersEndpoints = deprioritizedServers.Select(description => description.EndPoint);
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs:456]: minor: define `deprioritizedServersEndpoints `once. Do this one.

**Review by BorisDog — CHANGES_REQUESTED** [review]
Looks good, few minor comments.

File: src/MongoDB.Driver.Core/Core/Clusters/ICluster.cs [hunk 9]
```diff
@@ -81,20 +82,22 @@ public interface ICluster : IDisposable
...
         /// <summary>
-        /// Selects a server from the cluster.
+        /// Selects a server from the cluster and takes an optional collection of servers for deprioritization.
         /// </summary>
         /// <param name="selector">The server selector.</param>
         /// <param name="cancellationToken">The cancellation token.</param>
+        /// <param name="deprioritizedServers">The deprioritized Servers.</param>
         /// <returns>The selected server.</returns>
-        IServer SelectServer(IServerSelector selector, CancellationToken cancellationToken);
+        IServer SelectServer(IServerSelector selector, CancellationToken cancellationToken, IReadOnlyCollection<ServerDescription> deprioritizedServers = null);
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Clusters/ICluster.cs:91]: Consider having augmenting `IServerSelector` to account for `deprioritizedServers`.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Driver.Core/Core/Bindings/IBinding.cs (hunk 2 above)
Comment (adelinowona) [reply src/MongoDB.Driver.Core/Core/Bindings/IBinding.cs:72]: We settled with using the overloads

**Review by BorisDog — CHANGES_REQUESTED** [review]
Some minor comments.

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs [hunk 10]
```diff
@@ -413,6 +413,96 @@ public void SelectServer_should_throw_if_any_servers_are_incompatible(int min, i
...
+                _capturedEvents.Next().Should().BeOfType<ClusterSelectedServerEvent>();
+                _capturedEvents.Any().Should().BeFalse();
+            }
+        }
+
+        [Theory]
+        [ParameterAttributeData]
+        public void SelectServer_should_return_deprioritized_servers_if_no_other_servers_exist_or_cluster_not_sharded(
+            [Values(false, true)] bool async,
+            [Values(false, true)] bool IsSharded)
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs:466]: lowercase `isSharded`

File: tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs [hunk 11]
```diff
@@ -413,6 +413,93 @@ public void SelectServer_should_throw_if_any_servers_are_incompatible(int min, i
...
+
+                _capturedEvents.Next().Should().BeOfType<ClusterSelectingServerEvent>();
+                _capturedEvents.Next().Should().BeOfType<ClusterSelectedServerEvent>();
+                _capturedEvents.Any().Should().BeFalse();
+            }
+        }
+
+        [Theory]
+        [ParameterAttributeData]
+        public async void SelectServer_should_return_deprioritized_servers_if_no_other_servers_exist_or_cluster_not_sharded(
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs:464]: Forgot to mention that async method should return `Task`, for better clarity.

File: src/MongoDB.Driver.Core/Core/Operations/RetryableWriteOperationExecutor.cs [hunk 12]
```diff
@@ -53,7 +53,7 @@ public static TResult Execute<TResult>(IRetryableWriteOperation<TResult> operati
 
             try
             {
-                context.ReplaceChannelSource(context.Binding.GetWriteChannelSource(cancellationToken));
+                context.ReplaceChannelSource(context.Binding.GetWriteChannelSource(new [] { context.ChannelSource.ServerDescription }, cancellationToken));
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Operations/RetryableWriteOperationExecutor.cs:56]: Remove the extra space in `new []` --> `new[]`

File: src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+using System.Linq;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.Servers;
+
+namespace MongoDB.Driver.Core.Clusters.ServerSelectors
+{
+    /// <summary>
+    /// Represents a server selector that selects servers based on a collection of servers to deprioritize.
+    /// </summary>
+    public sealed class PriorityServerSelector : IServerSelector
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs:26]: We probably should have few very basic tests for this selector. like `EndPointServerSelectorTests`

File: src/MongoDB.Driver.Core/Core/Bindings/WritableServerBinding.cs [hunk 14]
```diff
@@ -62,58 +63,103 @@ public ICoreSessionHandle Session
...
                 throw new InvalidOperationException($"This overload of {nameof(GetWriteChannelSource)} cannot be called when pinned to a server.");
             }
 
-            var selector = new WritableServerSelector(mayUseSecondary);
+            var writableServerSelector = new WritableServerSelector(mayUseSecondary);
+
+            var selector = deprioritizedServers != null
+                ? (IServerSelector)new CompositeServerSelector(new IServerSelector[] { new PriorityServerSelector(deprioritizedServers), writableServerSelector })
+                : writableServerSelector;
+
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Bindings/WritableServerBinding.cs:162]: `new IServerSelector[] `-> `new[]`

File: src/MongoDB.Driver.Core/Core/Bindings/WritableServerBinding.cs [hunk 15]
```diff
@@ -62,58 +63,103 @@ public ICoreSessionHandle Session
...
             ThrowIfDisposed();
-            var server = await _cluster.SelectServerAndPinIfNeededAsync(_session, WritableServerSelector.Instance, cancellationToken).ConfigureAwait(false);
+            var server = await _cluster.SelectServerAndPinIfNeededAsync(_session, WritableServerSelector.Instance, deprioritizedServers, cancellationToken).ConfigureAwait(false);
             return CreateServerChannelSource(server);
         }
 
         /// <inheritdoc/>
         public async Task<IChannelSourceHandle> GetWriteChannelSourceAsync(IMayUseSecondaryCriteria mayUseSecondary, CancellationToken cancellationToken)
+        {
+            return await GetWriteChannelSourceAsync(null, mayUseSecondary, cancellationToken).ConfigureAwait(false);
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Bindings/WritableServerBinding.cs:146]: No need to await, can just return the task.
Additional await might create addition wrapping task object.

File: src/MongoDB.Driver.Core/Core/Bindings/WritableServerBinding.cs [hunk 16]
```diff
@@ -62,58 +63,103 @@ public ICoreSessionHandle Session
...
             if (IsSessionPinnedToServer())
             {
                 throw new InvalidOperationException($"This overload of {nameof(GetWriteChannelSource)} cannot be called when pinned to a server.");
             }
 
-            var selector = new WritableServerSelector(mayUseSecondary);
+            var writableServerSelector = new WritableServerSelector(mayUseSecondary);
+
+            var selector = deprioritizedServers != null
+                ? (IServerSelector)new CompositeServerSelector(new IServerSelector[] { new PriorityServerSelector(deprioritizedServers), writableServerSelector })
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Bindings/WritableServerBinding.cs:122]: Same [comment](https://github.com/mongodb/mongo-csharp-driver/pull/1304/files#r1612347781) applies here

File: src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        private readonly IReadOnlyCollection<ServerDescription> _deprioritizedServers;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PriorityServerSelector"/> class.
+        /// </summary>
+        /// <param name="deprioritizedServers">The collection of servers to deprioritize.</param>
+        public PriorityServerSelector(IReadOnlyCollection<ServerDescription> deprioritizedServers)
+        {
+            _deprioritizedServers = Ensure.IsNotNull(deprioritizedServers, nameof(deprioritizedServers));
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs:36]: Should this be `IsNotNullOrEmpty` maybe?

File: src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        private readonly IReadOnlyCollection<ServerDescription> _deprioritizedServers;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PriorityServerSelector"/> class.
+        /// </summary>
+        /// <param name="deprioritizedServers">The collection of servers to deprioritize.</param>
+        public PriorityServerSelector(IReadOnlyCollection<ServerDescription> deprioritizedServers)
+        {
+            _deprioritizedServers = Ensure.IsNotNullOrEmpty(deprioritizedServers, nameof(deprioritizedServers)) as IReadOnlyCollection<ServerDescription>;
```
Comment (BorisDog) [inline src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs:36]: I wish we had appropriate overload, to skip the `as` .

Comment (JamesKovacs) [reply src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs:36]: Why can't our list of deprioritized servers be empty? Wouldn't this be the case on the first call and wouldn't it only contain servers on the subsequent retry?

Right now 26 out of 28 patch variants are failing because `deprioritizedServers` is empty.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Looks like the logic around `PriorityServerSelector` isn't quite right as we have a lot of failing variants complaining about `deprioritizedServers` being empty. Please investigate and resolve. Thanks.

File: src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs (hunk 18 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Clusters/ServerSelectors/PriorityServerSelector.cs:36]: `PriorityServerSelector` is created conditionally (`deprioritizedServers != null`), so extending this condition to involve `PriorityServerSelector` only when deprioritizedServers is not empty seems natural.

**Review by JamesKovacs — APPROVED** [review]
LGTM
